DOCSP-37554 Retrieve Distinct Values (#61)
mongodb/docs-pymongo: source/read/distinct.txt

.. _pymongo-distinct:

==============================
Retrieve Distinct Field Values
==============================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: read, unique, code example

Overview
--------

> In this guide, you can learn how to retrieve a list of distinct values across a collection for a
> specified field with {+driver-short+}.

Reviewer: Thought: Would this be easier to parse as multiple sentences? Something like:

```suggestion
Within a collection, different documents might contain different values for a single field. 
For example, one ``restaurant`` document has a ``borough`` value of ``"Manhattan"``, and
another has a ``borough`` of ``"Queens"``.

With {+driver-short+}, you can retrieve all the distinct values
that a field contains across multiple documents.
```

If you like this format, you could do something similar in the subsections, like "We need to find all Italian restaurants in Manhattan" or something

Author: Yeah I like that, I was struggling to find a good way to phrase it in a short concise sentence. I changed the overview section to this and left the subsections as-is

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``sample_restaurants.restaurants`` collection
from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:ref:`<pymongo-get-started>`.

``distinct()`` Method
---------------------

To retrieve the distinct values for a specified field, call the ``distinct()``
method and pass in the name of the field you want to find distinct values for.

Retrieve Distinct Values Across a Collection
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example retrieves the distinct values of the ``borough`` field in
the ``restaurants`` collection:

.. io-code-block::

   .. input:: /includes/read/distinct.py
      :start-after: start-distinct
      :end-before: end-distinct
      :language:  python

   .. output:: 

      <placeholder output>

The results show each distinct value that appears in the ``borough`` field
across all documents in the collection. Although several documents have the same

> value in the ``borough`` field, each value only appears in the results once.

Reviewer: ```suggestion
value in the ``borough`` field, each value appears in the results only once.
```

You can provide a **query filter** to the ``distinct()`` method to find the distinct
field values across a subset of documents in a collection. A query filter is an expression that specifies search
criteria used to match documents in an operation. For more information about
creating a query filter, see :ref:`pymongo-specify-query`.

The following example retrieves the distinct values of the ``borough`` field for
all documents that have a ``cuisine`` field value of ``"Italian"``:

.. io-code-block::

   .. input:: /includes/read/distinct.py
      :start-after: start-distinct-with-query
      :end-before: end-distinct-with-query
      :language:  python

   .. output:: 

      <placeholder output>

Modify Distinct Behavior
~~~~~~~~~~~~~~~~~~~~~~~~

The ``distinct()`` method accepts optional parameters, which represent
options you can use to configure the operation. If you don't specify any
options, the driver does not customize the operation.

The following table describes the options you can set to customize
``distinct()``:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Property
     - Description

   * - ``filter``

>      - | A query filter that specifies which documents to retrieve distinct

Reviewer: ```suggestion
     - | A query filter that specifies the documents to retrieve distinct
```
